ProcessOptimizer/tests/test_noise_model.py

```diff
@@ -0,0 +1,200 @@
+
+@pytest.mark.parametrize("input",(1,2,3))
+def test_data_dependent_noise(signal_list, input):
+    # noise_choice returns a noise model that gives the same noise as the input data.
+    def noise_choice(X):
+        local_noise_model = AdditiveNoise()
+        local_noise_model._noise_distribution = lambda: X
+        return local_noise_model
+    noise_model = DataDependentNoise(noise_models=noise_choice)
+    noise_list = [noise_model.get_noise(input,signal) for signal in signal_list]  
+    assert [noise==signal for (signal,noise) in zip(signal_list,noise_list)]
+
+def test_zero_noice(signal_list):
+    noise_model = ZeroNoise()
+    noise_list = [noise_model.get_noise(None,signal) for signal in signal_list]    
+    assert [noise==0 for noise in noise_list]
+
+# Testing that the examples in the docstring of DataDependentNoise work
+@pytest.mark.parametrize("magnitude",(1,2,3))
+def test_noise_model_example_1(long_signal_list, magnitude):
+    # the following two lines are taken from the docstring of DataDependentNoise
+    noise_choice = lambda X: AdditiveNoise(noise_size=X)
+    noise_model = DataDependentNoise(noise_models=noise_choice)
+    data = [magnitude]*len(long_signal_list)
+    noise_list = [noise_model.get_noise(x,signal) 
+                  for (x,signal) in zip(data,long_signal_list)]    
+    evaluate_random_dist(noise_list,magnitude)
+
+def test_noise_model_example_2(long_signal_list):
+    # the following two lines are taken from the docstring of DataDependentNoise
+    noise_choice = lambda X: ZeroNoise() if X[0]==0 else AdditiveNoise()
+    noise_model = DataDependentNoise(noise_models=noise_choice)
+    X=[0,10,5]
+    noise_list = [noise_model.get_noise(X,signal) for signal in long_signal_list]
+    (mean, spread) = norm.fit(noise_list)
+    # This yields the zero noise model, where the fitted parameters are exactly zero
+    assert mean==0
+    assert spread==0
+    X=[1,27,53.4]
+    noise_list = [noise_model.get_noise(X,signal) for signal in long_signal_list]
+    evaluate_random_dist(noise_list)
+
+def test_sum_noise(signal_list):
+    noise_model = SumNoise(noise_model_list=[
+        "additive",
+        {"model_type" : "multiplicative", "noise_size" : 1}])
+    noise_model.noise_model_list[0]._noise_distribution = lambda: 2
+    noise_model.noise_model_list[1]._noise_distribution = lambda: 3
+    noise_list = [noise_model.get_noise(None,signal) for signal in signal_list]
+    true_noise_list = [2 + 3 * signal for signal in signal_list]
+    assert noise_list == true_noise_list
+
+# raw_noise of SumNoise should not be accesible, since it is a compound NoiseModel, and
+# should ot have "its own" noise.
+def test_sum_noise_raw_noise_error():
+    noise_model = SumNoise(noise_model_list=[AdditiveNoise()])
+    with pytest.raises(TypeError):
+        noise_model.raw_noise
+
+# raw_noise of AdditiveNoise should not be accesible, since it is a compound NoiseModel,
+# and should ot have "its own" noise.
```

Comment:
Is the test here (test_data_dependent_raw_noise_error) a left-over from building the test of the SumNoise class? Because with the way you have built AdditiveNoise (ConstantNoise) now, it should definitely have its own noise? Or am I misunderstanding what this test should do?